python/common/sql-versioning/sql_versioning/versioned_history.py

```diff
@@ -0,0 +1,256 @@
+"""Versioned mixin class and other utilities."""
+import datetime
+from contextlib import suppress
+
+from sqlalchemy import (BigInteger, Column, DateTime, Integer, SmallInteger,
+                        String, Table, and_, event, func, insert, inspect, or_,
+                        select, update)
+from sqlalchemy.ext.declarative import declarative_base, declared_attr
+from sqlalchemy.orm import Session, mapper
+
+Base = declarative_base()
+
+class Transaction(Base):
+    __tablename__ = 'transaction'
+
+    id = Column(BigInteger, primary_key=True, autoincrement=True)
+    issued_at = Column(DateTime(timezone=False), default=datetime.datetime.utcnow, nullable=True)
+    remote_addr = Column(String(50), nullable=True)
+
+    def __repr__(self):
+        return f"<Transaction(id={self.id}, issued_at={self.issued_at})>"
+
+
+class TransactionManager:
+
+    def __init__(self, session):
+        self.session = session
+
+    def create_transaction(self):
+        print("Entering create_transaction")
+        if 'current_transaction_id' in self.session.info:
+            print(f"Reusing existing transaction: {self.session.info['current_transaction_id']}")
+            return self.session.info['current_transaction_id']
+
+        # Use insert().returning() to get the ID and issued_at without committing
+        stmt = insert(Transaction).values(
+            issued_at = None
+        ).returning(Transaction.id, Transaction.issued_at)
+        result = self.session.execute(stmt)
+        transaction_id, issued_at = result.first()
+
+        print(f"Created new transaction: {transaction_id}")
+
+        self.session.info['current_transaction_id'] = transaction_id
+        print(f"Set current_transaction_id: {transaction_id}")
+        return transaction_id
+
+    def get_current_transaction_id(self):
+        return self.session.info.get('current_transaction_id')
+
+    def clear_current_transaction(self):
+        print(f"Clearing current transaction: {self.session.info.get('current_transaction_id')}")
+        self.session.info.pop('current_transaction_id', None)
+
+
+class Versioned:
+
+    is_enable = None  # debugging
+
+    @declared_attr
+    def __versioned_cls__(cls):
+        return cls.get_or_create_version_class()
+
+    @classmethod
+    def get_or_create_version_class(cls):
+        if not hasattr(cls, '_version_cls'):
+            class_name = f'{cls.__name__}Version'
+            table_name = f'{cls.__tablename__}_version'
+
+            attrs = {
+                '__tablename__': table_name,
+                'id': Column(Integer, primary_key=True),
+                'transaction_id': Column(BigInteger, primary_key=True, nullable=False),
+                'end_transaction_id': Column(BigInteger, nullable=True),
+                'operation_type': Column(SmallInteger, nullable=False),
```

Comment:
Should use a composite primary key (`id` + `transaction_id`). The types used are also consistent with those in `continuum `.